Respx.mock(...) decorator on test bug

Hi! You did a nice job with refactoring, I really like it! 

I found a bug and I have no idea how to fix it correctly. 
```python
@respx.mock(base_url='https://foo.bar')
def test_base_url_with_decorator(respx_mock: MockTransport):
    respx_mock.get('buzz')

    with httpx.Client() as client:
        client.get('https://foo.bar/buzz')

    respx_mock.calls.assert_called_once()
```

If you try to execute it you'll get an error:
```
  @respx.mock(base_url='https://foo.bar')
  def test_base_url_with_decorator(respx_mock: MockTransport):
E       fixture 'respx_mock' not found
>       available fixtures: cache, capfd, capfdbinary, caplog, capsys, capsysbinary, client, cov, doctest_namespace, event_loop, mocked_foo, mocked_ham, monkeypatch, my_mock, no_cover, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, unused_tcp_port, unused_tcp_port_factory
>       use 'pytest --fixtures [testpath]' for help on them.
```

The only idea I have is to create respx_mock a fixture (it will allow using respx like `monkeypatch` fixture) and tie it with the decorator.

--- a/respx/router.py
+++ b/respx/router.py
@@ -1,5 +1,5 @@
 import inspect
-from functools import wraps
+from functools import update_wrapper
 from types import TracebackType
 from typing import Any, Callable, Dict, List, Optional, Tuple, Type, Union, overload
 from warnings import warn
@@ -330,7 +330,6 @@ class MockRouter(Router):
             return respx_mock
 
         # Async Decorator
-        @wraps(func)
         async def async_decorator(*args, **kwargs):
             assert func is not None
             if self._local:
@@ -339,13 +338,16 @@ class MockRouter(Router):
                 return await func(*args, **kwargs)
 
         # Sync Decorator
-        @wraps(func)
         def sync_decorator(*args, **kwargs):
             assert func is not None
             if self._local:
                 kwargs["respx_mock"] = self
             with self:
                 return func(*args, **kwargs)
+
+        if not self._local:
+            async_decorator = update_wrapper(async_decorator, func)
+            sync_decorator = update_wrapper(sync_decorator, func)
 
         # Dispatch async/sync decorator, depening on decorated function.
         # - Only stage when using global decorator `@respx.mock`

--- a/respx/router.py
+++ b/respx/router.py
@@ -277,7 +277,6 @@ class Router:
 class MockRouter(Router):
     Mock: Optional[Type[BaseMock]]
     handler = Router.resolve
-    _local = False
 
     def __init__(
         self,
@@ -326,13 +325,16 @@ class MockRouter(Router):
             if assert_all_mocked is not None:
                 settings["assert_all_mocked"] = assert_all_mocked
             respx_mock = self.__class__(**settings)
-            respx_mock._local = True
             return respx_mock
+
+        # Determine if decorated function needs a `respx_mock` instance
+        argspec = inspect.getfullargspec(func)
+        needs_mock_reference = "respx_mock" in argspec.args
 
         # Async Decorator
         async def async_decorator(*args, **kwargs):
             assert func is not None
-            if self._local:
+            if needs_mock_reference:
                 kwargs["respx_mock"] = self
             async with self:
                 return await func(*args, **kwargs)
@@ -340,12 +342,12 @@ class MockRouter(Router):
         # Sync Decorator
         def sync_decorator(*args, **kwargs):
             assert func is not None
-            if self._local:
+            if "respx_mock" in argspec.args:
                 kwargs["respx_mock"] = self
             with self:
                 return func(*args, **kwargs)
 
-        if not self._local:
+        if not needs_mock_reference:
             async_decorator = update_wrapper(async_decorator, func)
             sync_decorator = update_wrapper(sync_decorator, func)
 

--- a/tests/test_mock.py
+++ b/tests/test_mock.py
@@ -128,6 +128,35 @@ async def test_local_async_decorator(client):
     assert respx.calls.call_count == 0
 
 
+def test_local_decorator_with_reference():
+    router = respx.mock()
+
+    @router
+    def test(respx_mock):
+        assert respx_mock is router
+
+    test()
+
+
+def test_local_decorator_without_reference():
+    router = respx.mock()
+    route = router.get("https://foo.bar/") % 202
+
+    @router
+    def test():
+        assert respx.calls.call_count == 0
+        response = httpx.get("https://foo.bar/")
+        assert route.called is True
+        assert response.status_code == 202
+        assert respx.calls.call_count == 0
+        assert router.calls.call_count == 1
+
+    assert router.calls.call_count == 0
+    assert respx.calls.call_count == 0
+    test()
+    assert respx.calls.call_count == 0
+
+
 @pytest.mark.asyncio
 async def test_global_contextmanager(client):
     with respx.mock:

--- a/tests/test_mock.py
+++ b/tests/test_mock.py
@@ -287,7 +287,7 @@ async def test_configured_router_reuse(client):
 
 @pytest.mark.asyncio
 @respx.mock(base_url="https://ham.spam/")
-async def test_nested_base_url(respx_mock=None):
+async def test_nested_base_url(respx_mock):
     request = respx_mock.patch("/egg/") % dict(content="yolk")
     async with respx.mock(base_url="https://foo.bar/api/") as foobar_mock:
         request1 = foobar_mock.get("/baz/") % dict(content="baz")

